### PR TITLE
Improve Acoustic performance

### PR DESCRIPTION
**Pull request by ChrisZYJ**
## Description

Enhance performance when using acoustic sources by:

1. Pre-calculating the spatial components of source strengths before the main time loops
2. Skipping the calculations for spatial locations far from the acoustic source supports

The performance improvement increases as the resolution increases. For the 3D_acoustic_support11 example (99^3 resolution), there is a 96% reduction in computational time for acoustic subroutines ('cuda_gpu_kern_sum' stats report).

### Type of change

- [x] Performance improvement

### Scope

- [x] This PR comprises a set of related changes with a common goal

## How Has This Been Tested?

Passed all test suites locally on CPU and GPU. Performance reports attached. 

## Checklist

- [x] I have added comments for the new code
- [x] I added Doxygen docstrings to the new code
- [x] I have made corresponding changes to the documentation (`docs/`)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [x] I ran `./mfc.sh format` before committing my code
- [x] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled (only NVHPC)
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [x] Checked that the code compiles using NVHPC compilers
- [ ] Checked that the code compiles using CRAY compilers
- [ ] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [ ] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [x] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [x] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran an Omniperf profile using `./mfc.sh run XXXX --gpu -t simulation --omniperf`, and have attached the output file and plain text results to this PR.
- [ ] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

## Nsight report for 3D_acoustic_support11 example:

[nsys_old.txt](https://github.com/user-attachments/files/16408713/nsys_old.txt)
[nsys_new.txt](https://github.com/user-attachments/files/16408710/nsys_new.txt)

### CODE REVIEW

File: src/simulation/m_acoustic_src.fpp [hunk 1]
```diff
@@ -78,12 +81,17 @@ module m_acoustic_src
     !> @}
     !$acc declare create(mass_src, e_src, mom_src)
 
+    type(source_spatial_type), dimension(:), allocatable :: source_spatials !< Data of non-zero source grid points for each source
+    !$acc declare create(source_spatials)
+
 #endif
 
+    integer, dimension(:), allocatable :: source_spatials_num_points !< Number of non-zero source grid points for each source (CPU only to prevent GPU bug)
```
Comment (sbryngelson) [inline src/simulation/m_acoustic_src.fpp]: This doesn't seem right. `source_spatials_num_points` isn't declared on the device anywhere but it is an allocatable, so it should be.

**Comment by sbryngelson** [comment]
@ChrisZYJ how did you determine that the performance was bad? how bad? out of curiosity and possible improvement of our tests, not punishment 😄

File: src/simulation/m_acoustic_src.fpp (hunk 1 above)
Comment (ChrisZYJ) [reply src/simulation/m_acoustic_src.fpp]: I wanted to make it explicit that I'm only using `source_spatials_num_points` as a CPU variable. But you're right, it's probably this that's causing the error on Frontier. I'll change it

**Comment by ChrisZYJ** [comment]
> @ChrisZYJ how did you determine that the performance was bad? how bad? out of curiosity and possible improvement of our tests, not punishment 😄

@sbryngelson To clarify, my previous PR didn't alter the performance. The code has always looped through all grid points, including those far from source supports, at every time step. This PR aims to improve that inefficiency, unrelated to CI.

The acoustic subroutine has consistently used a significant portion of simulation time:
3D planar support (99^3): 4.6% of total time
3D 18-element transducer array (99^3): 29.0% of total time

This PR reduces the latter from 29.0% to 1.4%, which should be helpful for large 3D acoustic simulations.

**Comment by sbryngelson** [comment]
> > @ChrisZYJ how did you determine that the performance was bad? how bad? out of curiosity and possible improvement of our tests, not punishment 😄
> 
> @sbryngelson To clarify, my previous PR didn't alter the performance. The code has always looped through all grid points, including those far from source supports, at every time step. This PR aims to improve that inefficiency, unrelated to CI.
> 
> The acoustic subroutine has consistently used a significant portion of simulation time:
> 
> * 3D planar support (99^3): 4.6% of total time
> * 3D 18-element transducer array (99^3): 29.0% of total time
> 
> This PR reduces the latter from 29.0% to 1.4%, which should be helpful for large 3D acoustic simulations.

Well done! 👏

**Comment by ChrisZYJ** [comment]
@sbryngelson It failed the Frontier test again on the acoustic cases, and it said it cannot execute MFC. Since I'm unable to run the case directly on Frontier, I'm not sure what the execution error is. Do you have any ideas on how to fix it? Thank you!

**Comment by sbryngelson** [comment]
rerunning it. the error was 
```
 Syscheck: PASSED.

mfc: OK > :) Running pre_process:

 Pre-processing a 299x0x0 case on 1 rank(s)
 Processing patch 1
 Processing patch 2
 Processing patch 3
 Elapsed Time 3.41100000000000084E-3

mfc: OK > :) Running simulation:

 Simulating a regular 299x0x0 case on 1 rank(s) with OpenACC offloading.

mfc: ERROR > :( /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/c4aa7d20bc/bin/simulation failed with exit code 1.

 

Error: Submitting batch file for Interactive failed. It can be found here: /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/1B0A7157/MFC.sh. Please check the file for errors.


mfc: ERROR > main.py finished with a 143 exit code.
mfc: (venv) Exiting the Python virtual environment.

   Failed test tests/1B0A7157: 1D -> Acoustic Source -> Sine -> Frequency after 3 attempt(s).

 

Error: Test tests/1B0A7157: 1D -> Acoustic Source -> Sine -> Frequency: Failed to execute MFC.

./mfc.sh: line 51: 54626 Terminated              python3 "$(pwd)/toolchain/main.py" "$@"

mfc: ERROR > main.py finished with a 143 exit code.
mfc: (venv) Exiting the Python virtual environment.
```
which is a bit strange.

**Comment by ChrisZYJ** [comment]
> rerunning it. the error was
> 
> ```
>  Syscheck: PASSED.
> 
> mfc: OK > :) Running pre_process:
> 
>  Pre-processing a 299x0x0 case on 1 rank(s)
>  Processing patch 1
>  Processing patch 2
>  Processing patch 3
>  Elapsed Time 3.41100000000000084E-3
> 
> mfc: OK > :) Running simulation:
> 
>  Simulating a regular 299x0x0 case on 1 rank(s) with OpenACC offloading.
> 
> mfc: ERROR > :( /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/c4aa7d20bc/bin/simulation failed with exit code 1.
> 
>  
> 
> Error: Submitting batch file for Interactive failed. It can be found here: /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/1B0A7157/MFC.sh. Please check the file for errors.
> 
> 
> mfc: ERROR > main.py finished with a 143 exit code.
> mfc: (venv) Exiting the Python virtual environment.
> 
>    Failed test tests/1B0A7157: 1D -> Acoustic Source -> Sine -> Frequency after 3 attempt(s).
> 
>  
> 
> Error: Test tests/1B0A7157: 1D -> Acoustic Source -> Sine -> Frequency: Failed to execute MFC.
> 
> ./mfc.sh: line 51: 54626 Terminated              python3 "$(pwd)/toolchain/main.py" "$@"
> 
> mfc: ERROR > main.py finished with a 143 exit code.
> mfc: (venv) Exiting the Python virtual environment.
> ```
> 
> which is a bit strange.

Thanks! I think I encountered similar errors before when running `./mfc.sh test`. When that happens, if I run that specific test case using `./mfc.sh run tests/[hash]/case.py`, it will output some useful messages. But I don't have access to Frontier, so I'm not sure how I can debug it

**Comment by sbryngelson** [comment]
You are correct that it is challenging to debug without Frontier access. I will see what I can do to help. it is re-running right now

**Comment by ChrisZYJ** [comment]
> You are correct that it is challenging to debug without Frontier access. I will see what I can do to help. it is re-running right now

Thank you so much!

**Comment by ChrisZYJ** [comment]
@sbryngelson Increasing tol from 2e-12 to 3e-12 passed the Frontier test! Sorry for the delay after fixing ACC_SETUP; I was confused by this message in the last failing commit (3540905), which led me to believe the issue was much more complex:

```Error: Test tests/3053B44F: 1D -> Acoustic Source -> Gaussian -> Sigma Dist: Variable n°302 (1-indexed) in D/prim.2.00.000050.dat is not within tolerance:
  - Candidate:   0.00541409128719
  - Golden:      0.00541409128954
  - Error:       abs: 2.35E-12, rel: 4.34E-10
  - Tolerance:   abs: 2.00E-12, rel: 2.00E-12
```

This suggested a significant increase in relative error. However, setting tol to 3e-12 passed the test. 

It leaves me very curious: Does the line 

```            err, msg = packtol.compare(pack, packer.load(golden_filepath), packtol.Tolerance(tol, tol))```

not check for relative tolerance? Thank you!

**Comment by sbryngelson** [comment]
> @sbryngelson Increasing tol from 2e-12 to 3e-12 passed the Frontier test! Sorry for the delay after fixing ACC_SETUP; I was confused by this message in the last failing commit ([3540905](https://github.com/MFlowCode/MFC/commit/3540905584acca094c35a79a45cceb17fc684de7)), which led me to believe the issue was much more complex:
> 
> ```
>   - Candidate:   0.00541409128719
>   - Golden:      0.00541409128954
>   - Error:       abs: 2.35E-12, rel: 4.34E-10
>   - Tolerance:   abs: 2.00E-12, rel: 2.00E-12
> ```
> 
> This suggested a significant increase in relative error. However, setting tol to 3e-12 passed the test.
> 
> It leaves me very curious: Does the line
> 
> ` err, msg = packtol.compare(pack, packer.load(golden_filepath), packtol.Tolerance(tol, tol))`
> 
> not check for relative tolerance? Thank you!

I'm not sure, what does the toolchain code suggest?

**Comment by henryleberre** [comment]
@ChrisZYJ The toolchain actually only requires either the absolute or the relative error to be within tolerance. Otherwise spurious issues would arise when we divide by small numbers and such. Apologies for the confusion, feel free to modify the console output to supply this bit of information.

**Comment by ChrisZYJ** [comment]
Thanks for the clarification Henry! This makes much more sense now. I can make that message clearer

**Comment by ChrisZYJ** [comment]
I think the PR is ready. Sorry for the wait!

**Review by sbryngelson — APPROVED** [review]

**Pull request merged**